### PR TITLE
Add a check that PIO env set to one of the options in pins.h

### PR DESCRIPTION
**Pull request by ellensp**
### Description

NB much code cloned from buildroot/share/vscode/auto_build.py

Add a routine check_suitable_env to common-dependencies.py 
The routine checks that the selected build environment (either in paltformio or via command line) 
is a valid option from the list of env's in pins.h for the selected motherboard.

EG if you select MOTHERBOARD BOARD_RAMBO, but leave the build env set to mega2560, this will build, but will not work correctly. With  check_suitable_env PlatformIO will now stop with the error
`You have MOTHERBOARD set to BOARD_RAMBO, please use default_envs = rambo
`

### Requirements

A invalid MOTHERBOARD and default_envs combination 

### Benefits

Less confused users

### Related Issues
Invalid bug report due to user error Issue: https://github.com/MarlinFirmware/Marlin/issues/21000

### CODE REVIEW

**Comment by ellensp** [comment]
CI test fails as "Run All Tests (mega1280)" is using MOTHERBOARD BOARD_TT_OSCAR which is listed as env:mega2560 only. 
Its is functioning as  intended to catch such miss matches.

**Comment by ellensp** [comment]
How ever this is in error...    
```
[Test STM32F103VE_longer] CLASSIC_UI U20 config...
You have MOTHERBOARD set to BOARD_LONGER3D_LK, please use default_envs = mega2560
```

**Comment by kad** [comment]
Parsing Configuration.h with just simple string marching is not the best idea. It will be start breaking on a bit more complex build variants and configs.

e.g. in my set of configs I have snippet like this:
https://github.com/kad/Marlin/blob/anycubic-kad-bugfix-2.0.x/Marlin/Configuration.h#L159-L169
```C
// Choose the name from boards.h that matches your setup
#ifndef MOTHERBOARD
  #if ENABLED(KAD_SKR_MINI)
    #define MOTHERBOARD BOARD_BTT_SKR_MINI_MZ_V1_0
  #elif ENABLED(KAD_SKR_E3_TURBO)
    #define MOTHERBOARD BOARD_BTT_SKR_E3_TURBO
  #else
    #define MOTHERBOARD BOARD_MELZI_CREALITY
  #endif
#endif
```
this allows me to have single config file for multiple custom build targets in platformio.

**Comment by ellensp** [comment]
checking for (board_name)broke some edge cases eg #elif MB(RAMPS_14_RE_ARM_EFB, RAMPS_14_RE_ARM_EEB...

**Comment by ellensp** [comment]
@kad "add check for multiple defined MOTHERBOARD and abort test"  should fix that, you just don't get the new test

File: buildroot/share/PlatformIO/scripts/common-dependencies.py [hunk 1]
```diff
@@ -322,6 +322,128 @@ def check_configfile_locations():
 				err = 'ERROR: Config files found in directory ' + str(p) + '. Please move them into the Marlin subdirectory.'
 				raise SystemExit(err)
 
+# Get the board being built from the Configuration.h file
+#   return: board name
+def get_board_name():
```
Comment (kad) [inline buildroot/share/PlatformIO/scripts/common-dependencies.py]: All of this can be replaced by using parsed  `env['MARLIN_FEATURES']['MOTHERBOARD']`

Comment (ellensp) [reply buildroot/share/PlatformIO/scripts/common-dependencies.py]: I looked but didn't see that...    using.

**Comment by ellensp** [comment]
Ok it passes all tests, but currently two tests are bypassed as they are non compliant 
```
Temporarily add env:mega1280 to MB(TT_OSCAR) just to skip that test
Temporarily add env:STM32F103RE_btt_USB to MB(BTT_SKR_CR6) just to skip that test. 
```
These need to be resolved properly.

**Comment by kad** [comment]
To give a hint on how python code can be written more efficiently, here is small example on how to get environments for all boards defined in Marlin source code:

```python
#!/usr/bin/env python3
import re

def get_starting_env(board_name_full):
    mb = board_name_full[6:]
    with open("Marlin/src/pins/pins.h","r") as f:
        mb_found = ""
        r=re.compile(r"if\s+MB\((.+)\)")
        for line in f.readlines():
            mbs = r.findall(line)
            if mbs:
                mb_found = mb if mb in re.split(",\s*", mbs[0]) else ""
            if mb_found and "#include " in line and "env:" in line:
                envs = [ e[4:] for e in re.findall("env:\w+", line) if e.startswith("env:") ]
                return envs
    return []

with open("Marlin/src/core/boards.h", "r") as f:
    for line in f.readlines():
        if "#define BOARD_" not in line:
            continue
        board = re.findall("BOARD_\w+", line)[0]
        print(board)
        print(get_starting_env(board))
```

However, my earlier comment is still valid: this script change will be failing if user defines own custom platformio environments.
e.g.:

```
# Base build for BTT SKR E3 Turbo
[env:anycubic_mega_zero_btt_e3turbo]
extends         = env:LPC1769
extra_scripts   = ${env:LPC1769.extra_scripts} pre:buildroot/share/PlatformIO/scripts/custom-binary-name.py
```

**Comment by ellensp** [comment]
If user is smart enough to add their own environments, they should be smart enough to add it to the list of valid environments in pins.h

**Comment by thisiskeithb** [comment]
I know we officially still support Arduino IDE (though I'm not sure how many Windows users are compiling Marlin with it successfully these days), but... we can probably dump all the environment checks at the top of the individual pins files as well since this is a more informative error message.

Example:
https://github.com/MarlinFirmware/Marlin/blob/52c246ae19549b5d99cd9a8e26cd7c8a1f0458b2/Marlin/src/pins/stm32f1/pins_BTT_SKR_MINI_E3_common.h#L24-L26

**Comment by kad** [comment]
> If user is smart enough to add their own environments, they should be smart enough to add it to the list of valid environments in pins.h

That is actually wrong attitude. Modifying `Configuration*.h` files and `platformio.ini` is one thing - that is a build config.
Modifying `pins.h` is already a **code** modification which is much harder to track and forget on rebases.

**Comment by ellensp** [comment]
@kad
If you create a new build environment, you must add it to the pins.h or the Marlin Auto build  will never use that environment.
So this is already a requirement to make Marlin user friendly.

**Comment by kad** [comment]
> If you create a new build environment, you must add it to the pins.h or the Marlin Auto build will never use that environment.

Marlin Auto Build is one of several ways of building source code. It is not "must", but more "good to have if you want to support Marlin Auto Build". To support one, new changes should not break other methods.

**Comment by ellensp** [comment]
How about if the ENV name contains the word 'custom' the test is skipped.  Would that make you happy?
You could add all the custom env's you like without issues.

**Comment by kad** [comment]
> How about if the ENV name contains the word 'custom' the test is skipped. Would that make you happy?
> You could add all the custom env's you like without issues.

That would be road to more problems. Proper way would be to have algorithm that would:
1. read platformio.ini, using ConfigParser (https://docs.python.org/3/library/configparser.html)
2. if `env['PIOENV']` not in list of sections() => ignore, as this is done via extensions to platformio.ini.
3. if `env['PIOENV']` does not match one of items from `pins.h`, but it has 'extends` -> follow that extends statement and check again (step 3) recursively. 

Overall about PR: 
- parsing files, comment above
- don't change `common-dependencies.py`, add your check as separate check script, that can be added to the list of `extra_scripts` after `common-cxxflags.py`  in `[common]` section of `platformio.ini`. That would eliminate problems of calling `load_marlin_features()` too early.

**Comment by ellensp** [comment]
I give up..

**Comment by rhapsodyv** [comment]
@ellensp and if you just add a warning message instead of an error? It may help a lot!!
